valhalla/lib/valhalla/storage/file_typed_disk.rb

```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+module Valhalla::Storage
+  class FileTypedDisk < Valkyrie::Storage::Disk
+    def initialize(master_storage_adapter:, derivative_storage_adapter:)
+      @master_storage_adapter = master_storage_adapter
+      @derivative_storage_adapter = derivative_storage_adapter
+    end
+
+    def upload(file:, resource: nil)
+      file_typed_adapter(file: file).upload(file: file, resource: resource)
+    end
+
+    private
+
+      def file_typed_adapter(file:)
+        file.derivative? ? @derivative_storage_adapter : @master_storage_adapter
```

Comment:
I don't think this'll work in the case of replacing derivatives from a file upload in the FileSet controller. Can't we just check the adapter for the previously existing file in that controller action and use that to upload?

Reply:
I'm uncertain as to how this could be supported.  `@change_set` is initialized from the PATCH parameters passed to the Controller (via `resource_params`):
```ruby
    def update
      @change_set = change_set_class.new(find_resource(params[:id])).prepopulate!
      authorize! :update, @change_set.resource
      if @change_set.validate(resource_params)
        @change_set.sync
        obj = nil
        change_set_persister.buffer_into_index do |persist|
          obj = persist.save(change_set: @change_set)
        end
```
(https://github.com/pulibrary/figgy/blob/master/valhalla/app/controllers/concerns/valhalla/resource_controller.rb#L62)

```ruby
    def resource_params
      params[resource_class.to_s.underscore.to_sym].to_unsafe_h
    end
```
(https://github.com/pulibrary/figgy/blob/master/valhalla/app/controllers/concerns/valhalla/resource_controller.rb#L90)

Example:
```bash
{"files"=>[{"473af163-d92c-4cb6-a036-63c0ce80a734"=>#<ActionDispatch::Http::UploadedFile:0x007fdc0aed0dd0 @tempfile=#<Tempfile:/var/folders/06/9nbfv2tx74jc0flpnq_mgz580000gp/T/RackMultipart20170927-472-4keg3b.jp2>, @original_filename="relax.jp2", @content_type="image/jp2", @headers="Content-Disposition: form-data; name=\"file_set[files[][473af163-d92c-4cb6-a036-63c0ce80a734]]\"; filename=\"test.jp2\"\r\nContent-Type: image/jp2\r\n">}]}
```

Because users can submit both master and derivative files into the same `FileSetChangeSet`, I would need to initialize a different ChangeSet for each of these files, with a separate storage adapter passed to new `PlumChangeSetPersister`'s used for each of these ChangeSets.